### ISSUE TITLE
How update session?

### ISSUE THREAD

**Issue by aloky** [sidebase/nuxt-auth#483]
### Environment

How update session? In next-auth `useSession()` return `update` method, but `useAuth()` not have update method...

### Describe the bug

If i change user.name example token use old(invalid) info.

**Comment by gritworks** [comment]
Need this too. update method is needed to update the session.

**Comment by asokr** [comment]
I solved it this way. But I don't know if this is the right approach
https://github.com/sidebase/nuxt-auth/discussions/474

**Comment by imduchy** [comment]
I'm also interested in this. I want to be able to refetch the user object. Or maybe there's any other recommended way to do this?

```ts
export default NuxtAuthHandler({
  // ...
  callbacks: {
    jwt: async ({ token, user }: any) => {
      // The user argument is only passed the first time this callback is called on a new session,
      // after the user signs in. In subsequent calls, only token will be available.
      if (user) {
        const userResponse = await $fetch("/api/auth/me");
        token.data = userResponse;
      }

      return Promise.resolve(token);
    },
  // ...
});
```

**Comment by aloky** [comment]
i solved it this way...

```javascript
export default NuxtAuthHandler({
  // ...
  callbacks: {
    jwt: async ({ token, trigger, user }) => {
      if (trigger === 'signIn') {
        token.user = user || await getMe({ email: token.email });
      } else {
        token.user = await getMe({ id: token.user.id });
      }

      if (!token.user) {
        return Promise.reject(token);
      }

      return Promise.resolve(token);
    },
  ...
```

**Comment by zoey-kaiser** [comment]
> i solved it this way...

In our in-house projects with `nuxt-auth` we solve it in a similar way. By connecting a prisma DB, we can make changes inside the Database and refetch the session. This will then pull the changes from our users table and ensure the data stays synced between the frontend and backend. 

I understand it makes sense that you may want to update the frontend session, but it will always require a certain amount of backend logic to also properly validate (as also mentioned by NextAuth in their docs)

**Comment by imduchy** [comment]
> > i solved it this way...
> 
> In our in-house projects with `nuxt-auth` we solve it in a similar way. By connecting a prisma DB, we can make changes inside the Database and refetch the session. This will then pull the changes from our users table and ensure the data stays synced between the frontend and backend.
> 
> I understand it makes sense that you may want to update the frontend session, but it will always require a certain amount of backend logic to also properly validate (as also mentioned by NextAuth in their docs)

@zoey-kaiser thank you for replying. Would you mind providing an example of how you guys do it @ sidebase?

An alternative approach that I ended up using is defining composables and leveraging built-in state management.

```ts
export const useAuthHelpers = () => {
  const { status, data } = useAuth();

  const isLoggedIn = computed(() => status.value === "authenticated");

  // Initially, the state is populated by fetching the user object from the `/auth/me` endpoint (also connected to a database)
  const userState = () =>
    useState<IUser | undefined>("user", () => {
      console.debug("(authHelpers.useState) Called");

      if (!isLoggedIn.value) {
        return undefined;
      }

      console.debug("(authHelpers.useState) Calling forceFetchUser");
      forceFetchUser();
    });

  // Whenever I update a user object, I call this function from within a component to re-fetch the object from the backend
  const forceFetchUser = async () => {
    console.debug("(authHelpers.forceFetchUser) Called");

    const _userState = useState<IUser | undefined>("user");
    _userState.value = await $fetch("/api/auth/me");
  };

// ...
}
```

However, I don't think this is the most elegant solution. My concern with the https://github.com/sidebase/nuxt-auth/issues/483#issuecomment-1685056727 solution is that the application would make a lot (excessive amount?) of API calls to the backend.

Thank you!

**Comment by zoey-kaiser** [comment]
> Would you mind providing an example of how you guys do it @ sidebase?

I sadly can't, as most of our code needs to stay internal. Ill see if I can make an anonymized snippet!

> My concern with the https://github.com/sidebase/nuxt-auth/issues/483#issuecomment-1685056727 solution is that the application would make a lot of (unnecessary?) API calls to the backend.

In our case this is not an issue, as any modification to the Prisma Database, returns the updated data anyway. This way, we would need to make the API call regardless of authstate, as the update method returns the new data anyway. 
In addition to this, new APIs calls are made regularly anyway, as a new one is made whenever the session needs to be re-fetched. If you correctly add caching to your backend, it should generally be robust enough to handle this (obviously until it reaches a certain scale). 

```ts
  const forceFetchUser = async () => {
    console.debug("(authHelpers.forceFetchUser) Called");

    const _userState = useState<IUser | undefined>("user");
    _userState.value = await $fetch("/api/auth/me");
  };
```

Based on your code example it also looks like you refetch your session whenever updating a value! You could also call the composable `const { getSession } = useAuth()` to achieve the same behaviour.

**Comment by imduchy** [comment]
> > Would you mind providing an example of how you guys do it @ sidebase?
> 
> I sadly can't, as most of our code needs to stay internal. Ill see if I can make an anonymized snippet!

Of course, I understand. If you managed to post a generic/anonymized snippet, I'm sure many developers would benefit from it 😉 

> Based on your code example it also looks like you refetch your session whenever updating a value! You could also call the composable `const { getSession } = useAuth()` to achieve the same behaviour.

I could definitely use `getSession` composable but at the moment, I removed any custom logic from the `jwt` and `session` callbacks. I noticed that the `jwt` callback got sometimes called multiple times (4x) on a route change—I didn't yet look into what was causing that.

So, would you suggest fetching a user object from the backend from within the `jwt` callback instead? And then, whenever I need to manually fetch the user, I'll simply call `getSession`.

**Comment by gritworks** [comment]
any progress on this? why is the update command not avalable? is this a technical or security limitation?

**Comment by anPalmer** [comment]
Any chance we could get support for the relatively recent Next Auth addition of update()? Being able to call it once, pass in an argument with data to use on the backend, means one call to the backend/DB instead of endless calls.
https://next-auth.js.org/getting-started/client#updating-the-session

**Comment by andreasvirkus** [comment]
+1 for providing a proper `update` method 🙏

**Comment by zoey-kaiser** [comment]
Hi everyone! I wanted to quickly give some more input from my side:

My proposal for how I would currently recommend updating the session is as follows: 
- User interacts with a UI element (say Avatar editing)
- User clicks on save
- A mutation is made on your backend
- After the mutation was successful you call `getSession` which updates all the session data inside of your app

What I really don't like about the `update` method implemented by NextAuth is that, as they warn it is client side only, with no server side validation. Without implementing proper server validation yourself, this could be a big risk, as a user could in theory eg. change their role, or other sensitive data. I feel like this can be a deep worm hole, which could become a bigger risk in your app, then if you just handled the user data updating separately and then afterwards refreshed the session. 

This would ensure that the data is never modified only client side and always goes through the server. IMO this is a much safer solution and also not that hard to implement. But your free to argue against my reasoning, or make proposals on how we could provide a `update` method, while ensuring it has a higher level of security then that of NextAuth.
